Bluesky Control Panel (Independent Publisher)

krautrocker opened the pull request:
---
### When submitting a connector, please make sure that you follow the requirements below, otherwise your PR might be rejected. We want to make you have a well-built connector, a smooth certification experience, and your users are happy :) 

If this is your first time submitting to GitHub and you need some help, please sign up for this [session](https://forms.office.com/pages/responsepage.aspx?id=KtIy2vgLW0SOgZbwvQuRaXDXyCl9DkBHq4A2OG7uLpdUMTFJWFFGVUxBNUFZQjZWRUdaOE5BMFkwNS4u). 

- [x] I attest that the connector doesn't exist on the Power Platform today. I've verified by checking the pull requests in GitHub and by searching for the connector on the platform or in the documentation. 
- [ ] I attest that the connector works and I verified by deploying and testing all the operations. 
- [x] I attest that I have added detailed descriptions for all operations and parameters in the swagger file.
- [x] I attest that I have added response schemas to my actions, unless the response schema is dynamic. 
- [x] I validated the swagger file, `apiDefinition.swagger.json`, by running `paconn validate` command.
- [x] If this is a certified connector, I confirm that `apiProperties.json` has a valid brand color and doesn't use an invalid brand color, `#007ee5` or `#ffffff`. If this is an independent publisher connector, I confirm that I am not submitting a connector icon.

If you are an Independent Publisher, you must also attest to the following to ensure a smooth publishing process:
- [x] I have named this PR after the pattern of "Connector Name (Independent Publisher)" ex: HubSpot Marketing (Independent Publisher)
- [x] Within this PR markdown file, I have pasted screenshots that show: 3 unique operations (actions/triggers) working within a Flow. This can be in one flow or part of multiple flows. For each one of those flows, I have pasted in screenshots of the Flow succeeding. 
- [x] Within this PR markdown file, I have pasted in a screenshot from the Test operations section within the Custom Connector UI.
- [x] If the connector uses OAuth, I have provided detailed steps on how to create an app in the readme.md.

krautrocker committed:
Initial commit - Bluesky Control Panel. All other non-app.bsky endpoints requiring their own connector separate from app.bsky

krautrocker committed:
Boop

krautrocker commented:
1.) Validation via paconn happens tomorrow. There's a few errors around schema and missing elements but nothing fatal. Clean-up.

2.) With the help of @troystaylor, we can confirm that two connectors are necessary - one for app.bsky endpoints and this connector for all other endpoints (ATProto, admin, chat, tools, if memory serves).

3.) For reference: https://github.com/bluesky-social/bsky-docs/blob/main/atproto-openapi-types/spec/api.json

Though this did require some reworking, such as removing "const" from all (70+) connectors, which brings the total up to over 100 when counting the app.bsky endpoints.

Feel free to contact me around contributing and maintaining. Lots going on here! =)

krautrocker commented:
@microsoft-github-policy-service agree company="Dan Romano aka IDR, Inc. dba SWOLCAT"

krautrocker committed:
- Refinement per UI validations

krautrocker committed:
- Swagger certified via paconn (but still has warnings)

krautrocker committed:
- Continued endpoint refinement per paconn validation (responses).

krautrocker committed:
- Added error response definitions

- Added additional endpoints using docs as reference

- Comprehensive but non-exhaustive list of endpoints and descriptions added to readme.md

vmanoharas commented:
Hello @krautrocker, 

Is this PR ready for review? if yes, kindly remove gitingore and checkin to the root folder. Also remove the Proposal word from the PR title. 

![image](https://github.com/user-attachments/assets/39351b31-7cdb-4be2-88d6-a0f6d119f3c7)

krautrocker committed:
Delete .gitignore

Delete .gitignore

krautrocker committed:
- Added additional endpoints

- Validated with paconn, no errors

krautrocker committed:
- Updated readme, WIP

krautrocker committed:
- Updated authentication

- Updated readme

- Validated with paconn, no errors

- Moved global definitions position

Ready to test with a client and secret.

krautrocker reviewed (commented):
Ready to test, but a few housekeeping items do remain. This PR is indeed different from this other [PR #3711](https://github.com/microsoft/PowerPlatformConnectors/pull/3706) but may merge the branches.

krautrocker committed:
- Authentication update, OAuth no dice

- Added connection parameters to apiProperties

- Added policy templates to apiProperties

- Validated with paconn, no errors

krautrocker commented:
Three successful tests:

1.) Create Session

![Screenshot 2024-11-18 211932](https://github.com/user-attachments/assets/c56c87e8-0de2-4c76-93b9-37dc922e11f8)

2.) Describe Server

![Screenshot 2024-11-18 213449](https://github.com/user-attachments/assets/d6a14fda-975f-4b86-a174-a01d012747ab)

3.) Check Account Status

![Screenshot 2024-11-18 212142](https://github.com/user-attachments/assets/330598a7-2776-4c94-8d35-4c66666684b5)

krautrocker commented:
Three actions working in a flow:

![Screenshot 2024-11-18 214405](https://github.com/user-attachments/assets/65e9b0f7-fae6-47b5-8dad-535d14b90498)

krautrocker committed:
- Fixed security definitions

krautrocker committed:
- Further update to apiProperties

krautrocker commented:
Reflecting on the past week of work. A heckuva lot has been done and I'm working to make connector definitions line up to the API definitions.

1.) Endpoints require QA.

I burned through these to get us on the beach. Now, I need to go through the catalogue and conduct a good QA. This will involve the Bluesky OpenAPI definition file, though the definition will need to be modified to work in the connector (components vs security definitions, for example).

Responses are present for success and failures. Defaults may need to be added to some endpoints.

2.) apiProperties feature connection parameters and policy templates.

Params and templates help us authenticate and operate consistently across the many endpoints. No repeats or manual intervention necessary by using these.

3.) Two connectors...or all-in-one?

The OpenAPI definition from Bluesky is quite large, but both the Bluesky Social and Control Panel connectors have global definitions (security, error responses, etc.), which reduces the size of the connector. Still, I'm not sure if all the endpoints can fit into one while not exceeding the size limit. If they don't breach the limit, then we have an opportunity for a single connector.

4.) Authentication

Some back and forth here on what works and what doesn't. OAuth abandoned for now and tokens have taken the lead. Query string is also a possibility, and though it may theoretically work, this doesn't seem ideal.

/**/

More tomorrow. Stay tuned.
